process/block/preprocess/transactionsV2.go

```diff
@@ -0,0 +1,416 @@
+)
+
+func (txs *transactions) createAndProcessMiniBlocksFromMeV2(
+	haveTime func() bool,
+	isShardStuck func(uint32) bool,
+	isMaxBlockSizeReached func(int, int) bool,
+	sortedTxs []*txcache.WrappedTransaction,
+) (block.MiniBlockSlice, error) {
+	log.Debug("createAndProcessMiniBlocksFromMeV2 has been started")
+
+	mapMiniBlocks := make(map[uint32]*block.MiniBlock)
+	mapSCTxs := make(map[string]struct{})
+	mapTxsForShard := make(map[uint32]int)
+	mapScsForShard := make(map[uint32]int)
+
+	numTxsAdded := 0
+	numTxsBad := 0
+	numTxsSkipped := 0
+	numTxsFailed := 0
+	numTxsWithInitialBalanceConsumed := 0
+	numCrossShardScCallsOrSpecialTxs := 0
+
+	totalTimeUsedForProcesss := time.Duration(0)
+	totalTimeUsedForComputeGasConsumed := time.Duration(0)
+
+	firstInvalidTxFound := false
+	firstCrossShardScCallOrSpecialTxFound := false
+
+	gasConsumedByMiniBlocksInSenderShard := uint64(0)
+	mapGasConsumedByMiniBlockInReceiverShard := make(map[uint32]uint64)
+	totalGasConsumedInSelfShard := txs.gasHandler.TotalGasConsumed()
+
+	log.Debug("createAndProcessMiniBlocksFromMeV2", "totalGasConsumedInSelfShard", totalGasConsumedInSelfShard)
+
+	senderAddressToSkip := []byte("")
+
+	defer func() {
+		go txs.notifyTransactionProviderIfNeeded()
+	}()
+
+	for shardID := uint32(0); shardID < txs.shardCoordinator.NumberOfShards(); shardID++ {
+		mapMiniBlocks[shardID] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), shardID, block.TxBlock)
+	}
+
+	mapMiniBlocks[core.MetachainShardId] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), core.MetachainShardId, block.TxBlock)
+
+	for index := range sortedTxs {
+		if !haveTime() {
+			log.Debug("time is out in createAndProcessMiniBlocksFromMeV2")
+			break
+		}
+
+		tx, ok := sortedTxs[index].Tx.(*transaction.Transaction)
+		if !ok {
+			log.Debug("wrong type assertion",
+				"hash", sortedTxs[index].TxHash,
+				"sender shard", sortedTxs[index].SenderShardID,
+				"receiver shard", sortedTxs[index].ReceiverShardID)
+			continue
+		}
+
+		txHash := sortedTxs[index].TxHash
+		senderShardID := sortedTxs[index].SenderShardID
+		receiverShardID := sortedTxs[index].ReceiverShardID
+
+		miniBlock, ok := mapMiniBlocks[receiverShardID]
+		if !ok {
+			log.Debug("miniblock is not created", "shard", receiverShardID)
+			continue
+		}
+
+		isReceiverSmartContractAddress := core.IsSmartContractAddress(tx.RcvAddr)
+		firstMiniBlockSplitForReceiverShardFound := (isReceiverSmartContractAddress && mapScsForShard[receiverShardID] == 0 && mapTxsForShard[receiverShardID] > 0) ||
+			(!isReceiverSmartContractAddress && mapTxsForShard[receiverShardID] == 0 && mapScsForShard[receiverShardID] > 0)
+
+		numNewMiniBlocks := 0
+		if len(miniBlock.TxHashes) == 0 || firstCrossShardScCallOrSpecialTxFound {
+			numNewMiniBlocks = 1
+		}
+		numNewTxs := 1
+
+		isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
+			(isReceiverSmartContractAddress || len(tx.RcvUserName) > 0)
```

Comment:
why would be a special tx if tx.RcvUsername > 0 ?

Reply:
This will protect the destination shard which could create many scr and the size of the resulted block could exceed the max limit of 1 MB, for sc calls and for txs with user name